File: day1.py
from aoc import *

def digitsFromNames(mixedStr, asChars=False):
  mixedStr = mixedStr.replace('one', 'o1e').replace('two', 't2o').replace('three', 't3e').replace('four', 'f4r').replace('five', 'f5e').replace('six', 's6x').replace('seven', 's7n').replace('eight', 'e8t').replace('nine', 'n9e')
  if asChars:
    return [val for val in mixedStr if val.isdigit()]
  else:
    return [int(val) for val in mixedStr if val.isdigit()]
  
class DayOne:

  mem = []
  vals = []
  valspt2 = []

  def __init__(self, mem):
    self.mem = mem.copy()
    for line in mem:
      self.vals.append(digits(line, True))
      self.valspt2.append(digitsFromNames(line, True))

  def partOne(self):
    return sum(int(i[0] + i[-1]) for i in self.vals)

  def partTwo(self):
    return sum(int(i[0] + i[-1]) for i in self.valspt2)

File: day1.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/phantom903/AoC2023/internal/aoc"
)

func part1(lines []string) int {
	acc := 0
	for _, line := range lines {
		digs := aoc.ExtractDigits(line)
		if len(digs) == 0 {
			continue
		}
		blob := fmt.Sprintf("%d%d", digs[0], digs[len(digs)-1])
		val, err := strconv.Atoi(blob)
		if err != nil {
			panic(fmt.Sprintf("Error converting string to int: %v", err))
		}
		acc += val
	}
	return acc
}

func part2(lines []string) int {
	acc := 0
	for _, line := range lines {
		replacer := strings.NewReplacer("one", "o1e", "two", "t2o", "three", "t3e", "four", "f4r", "five", "f5e", "six", "s6x", "seven", "s7n", "eight", "e8t", "nine", "n9e")
		line = replacer.Replace(line)
		digs := aoc.ExtractDigits(line)
		blob := fmt.Sprintf("%d%d", digs[0], digs[len(digs)-1])
		val, err := strconv.Atoi(blob)
		if err != nil {
			panic(fmt.Sprintf("Error converting string to int: %v", err))
		}
		acc += val
	}
	return acc
}

func main() {
	lines, err := aoc.FileOpenLines("input/day1.txt")

	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	fmt.Println("Part 1:", part1(lines))
	fmt.Println("Part 2:", part2(lines))
}
